File: tools/capture_frames.py
import sys
import os
import time
from playwright.sync_api import sync_playwright

def main():
    if len(sys.argv) < 6:
        print("Usage: python capture_frames.py <html_file> <frames_dir> <width> <height> <loop_ms>")
        sys.exit(1)
        
    html_file = os.path.abspath(sys.argv[1])
    frames_dir = os.path.abspath(sys.argv[2])
    width = int(sys.argv[3])
    height = int(sys.argv[4])
    loop_duration_ms = int(sys.argv[5])
    
    num_frames = 60
    frame_step = loop_duration_ms / num_frames
    
    os.makedirs(frames_dir, exist_ok=True)
    for f in os.listdir(frames_dir):
        if f.endswith('.png'):
            os.remove(os.path.join(frames_dir, f))
            
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page(
            viewport={"width": width, "height": height},
            device_scale_factor=2
        )
        
        file_url = f"file:///{html_file.replace(os.sep, '/')}"
        page.goto(file_url, wait_until="networkidle")
        time.sleep(1) # wait for fonts/setup
        
        print(f"Capturing {num_frames} frames...")
        
        for i in range(num_frames):
            t = 10000 + i * frame_step
            
            page.evaluate(f"""(ms) => {{
                document.getAnimations().forEach(anim => {{
                    if(anim.effect) {{
                        anim.currentTime = ms;
                        anim.pause();
                    }}
                }});
            }}""", t)
            
            frame_path = os.path.join(frames_dir, f"frame_{i:03d}.png")
            page.screenshot(path=frame_path)
            
            if (i+1) % 10 == 0:
                print(f"  Frame {i+1}/{num_frames}")
                
        browser.close()
        print("✅ Frames saved to:", frames_dir)

if __name__ == "__main__":
    main()

File: tools/capture_frames.js
const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs');

const args = process.argv.slice(2);
if (args.length < 5) {
    console.error("Usage: node capture_frames.js <html_file> <frames_dir> <width> <height> <loop_ms>");
    process.exit(1);
}

const HTML_FILE = path.resolve(args[0]);
const FRAMES_DIR = path.resolve(args[1]);
const WIDTH = parseInt(args[2]);
const HEIGHT = parseInt(args[3]);
const LOOP_DURATION_MS = parseInt(args[4]);
const NUM_FRAMES = 60; // Hardcoded to 60 for consistent ~5-second smooth loops
const FRAME_STEP = LOOP_DURATION_MS / NUM_FRAMES;

if (!fs.existsSync(FRAMES_DIR)) fs.mkdirSync(FRAMES_DIR, { recursive: true });
fs.readdirSync(FRAMES_DIR).filter(f => f.endsWith('.png')).forEach(f => fs.unlinkSync(path.join(FRAMES_DIR, f)));

(async () => {
    const browser = await puppeteer.launch({ headless: 'new' });
    const page = await browser.newPage();
    await page.setViewport({ width: WIDTH, height: HEIGHT, deviceScaleFactor: 2 });

    const fileUrl = `file:///${HTML_FILE.replace(/\\/g, '/')}`;
    await page.goto(fileUrl, { waitUntil: 'domcontentloaded' });

    // Wait for fonts to load
    await new Promise(r => setTimeout(r, 800));

    console.log(`Capturing ${NUM_FRAMES} frames...`);

    for (let i = 0; i < NUM_FRAMES; i++) {
        const t = 10000 + i * FRAME_STEP;

        await page.evaluate((ms) => {
            document.getAnimations().forEach(anim => {
                anim.currentTime = ms;
                anim.pause();
            });
        }, t);

        const framePath = path.join(FRAMES_DIR, `frame_${String(i).padStart(3, '0')}.png`);
        await page.screenshot({ path: framePath, type: 'png' });

        if (i % 10 === 0) console.log(`  Frame ${i + 1}/${NUM_FRAMES}`);
    }

    await browser.close();
    console.log('✅ Frames saved to:', FRAMES_DIR);
})();
